addons/sale/models/sale_order_line.py

```diff
@@ -564,33 +564,46 @@ def _get_pricelist_price_before_discount(self):
     @api.depends('product_id', 'product_uom', 'product_uom_qty')
     def _compute_discount(self):
         for line in self:
-            if not line.product_id or line.display_type:
-                line.discount = 0.0
+            line.discount = line._get_real_discount()
 
-            if not (
-                line.order_id.pricelist_id
-                and line.order_id.pricelist_id.discount_policy == 'without_discount'
-            ):
-                continue
+    def _get_real_discount(self, line_discount=None):
+        """ Get real pricelist discount without rounding.
+            Takes an optional parameter to compare the pricelist discount to.
+            Returns the given discount if it doesn't correspond to the
+            pricelist discount, i.e. when a discount was applied manually.
+        """
+        self.ensure_one()
 
-            line.discount = 0.0
+        if self.display_type or not self.product_id:
+            return 0.0
 
-            if not line.pricelist_item_id:
-                # No pricelist rule was found for the product
-                # therefore, the pricelist didn't apply any discount/change
-                # to the existing sales price.
-                continue
+        if not (
+            self.pricelist_item_id
+            and self.order_id.pricelist_id.discount_policy == 'without_discount'
```

Comment:
Check the pricelist before computing the applicable rule, because the rule computation is the costly part we want to avoid.

Also, maybe we could check whether there is a number in the max precision of the field before trying to recompute the discount ?

`6.09%` -> recompute
`6.40%` -> no need for recomputation

:thinking: